File: rest-service/api/src/main/java/org/veupathdb/service/multiblast/model/blast/p/BlastpTask.java
package org.veupathdb.service.multiblast.model.blast.p;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;

public enum BlastpTask
{
  BlastP("blastp"),
  BlastPFast("blastp-fast"),
  BlastPShort("blastp-short");

  private final String value;

  BlastpTask(String value) {
    this.value = value;
  }

  public String getValue() {
    return value;
  }

  @Override
  @JsonValue
  public String toString() {
    return value;
  }

  @JsonCreator
  public static BlastpTask fromString(String value) {
    for (var e : values())
      if (e.value.equals(value))
        return e;

    throw new IllegalArgumentException(); // TODO: print actual value
  }
}

File: lib-jvm-mblast-sdk/src/main/kotlin/org/veupathdb/lib/mblast/sdk/query/blast/BlastPTask.kt
package org.veupathdb.lib.mblast.sdk.query.blast

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonValue
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.node.TextNode
import org.veupathdb.lib.mblast.sdk.util.UnrecognizedEnumException
import org.veupathdb.lib.mblast.sdk.util.dEnumRequireTextual

enum class BlastPTask {
  BlastP,
  BlastPFast,
  BlastPShort,
  ;

  @JsonValue
  fun toJson(): JsonNode = TextNode(when (this) {
    BlastP      -> "blastp"
    BlastPFast  -> "blastp-fast"
    BlastPShort -> "blastp-short"
  })

  companion object {

    @JvmStatic
    @JsonCreator
    fun fromJson(js: JsonNode): BlastPTask {
      js.dEnumRequireTextual(BlastPTask::class)

      return when (val v = js.textValue()) {
        "blastp"       -> BlastP
        "blastp-fast"  -> BlastPFast
        "blastp-short" -> BlastPShort
        else           -> throw UnrecognizedEnumException(BlastPTask::class, v)
      }
    }
  }
}